server/middleware/oidc/index.js

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2024 Uber Technologies Inc.
+//
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+const OpenIDClient = require('openid-client');
+const passport = require('koa-passport');
+const session = require('koa-session');
+const allowUrl = ['/login', '/logout', '/oauth2/redirect'];
+
+let settings = {};
+
+try {
+  settings = require('../../config/oidc');
+} catch (e) {
+  console.log('OIDC configuration file not found, using ENV variables');
+}
+
+const clientID = settings.clientID || process.env.OPENID_CLIENT_ID;
+const clientSecret = settings.clientSecret || process.env.OPENID_CLIENT_SECRET;
+const callbackURL = settings.callbackURL || process.env.OPENID_CALLBACK_URL;
+const discoverURL = settings.discoverURL || process.env.OPENID_DISCOVER_URL;
+const scope = settings.scope || process.env.OPENID_SCOPE || 'openid';
+
+const middleware = async function(ctx, next) {
+  if (allowUrl.includes(ctx.path)) {
+    return next();
+  }
+
+  if (ctx.isUnauthenticated()) {
+    return ctx.redirect('/login');
+  }
+
+  // logout user to remove session with old token
+  if (
```

Comment:
Isn't this condition similar to the above?

Reply:
there's a situation when session is still valid, but token is expired. In this case expired token will be sent to server and user will see 401 errors. 

Not sure if configuring session TTL based on token TTL is a good idea.